### PR TITLE
Co\Redis compatibility mode

### PR DESCRIPTION
**Pull request by caohao-php**
repair co redis hGetAll

### CODE REVIEW

File: swoole_redis_coro.cc [hunk 1]
```diff
@@ -973,6 +974,39 @@ static sw_inline bool swoole_redis_coro_close(swRedisClient *redis)
...
 
+static sw_inline void swoole_redis_handle_result(zval* return_value, bool str2double) {
+    int index = 0;
+    zval *zkey, *zvalue;
+    zval zret;
+    array_init(&zret);
+        
+    ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(return_value), zvalue)
+    {
+        if(index == 0)
```
Comment (twose) [inline swoole_redis_coro.cc]:
```C
bool is_key = false;
if (!(is_key = !is_key))
{
    // xxxx
}
else
{
    // xxxx
}
```

File: swoole_redis_coro.cc [hunk 2]
```diff
@@ -973,6 +974,39 @@ static sw_inline bool swoole_redis_coro_close(swRedisClient *redis)
     return false;
 }
 
+static sw_inline void swoole_redis_handle_result(zval* return_value, bool str2double) {
```
Comment (twose) [inline swoole_redis_coro.cc]: is this can be named `swoole_redis_handle_assoc_array_result`?

File: swoole_redis_coro.cc [hunk 3]
```diff
@@ -3875,6 +3942,40 @@ static PHP_METHOD(swoole_redis_coro, hMGet)
...
+            zval zkey_str = *zkey;
+            zval_copy_ctor(&zkey_str);
+            convert_to_string(&zkey_str);
+            
+            zvalue = zend_hash_index_find(Z_ARRVAL_P(return_value), index++);
+            if(Z_ISNULL_P(zvalue))
+            {
+                zval zfalse;
+                ZVAL_FALSE(&zfalse);
+                zvalue = &zfalse;
```
Comment (twose) [inline swoole_redis_coro.cc]: memory error

File: swoole_redis_coro.cc [hunk 4]
```diff
@@ -3875,6 +3942,40 @@ static PHP_METHOD(swoole_redis_coro, hMGet)
...
+    if(redis->compatibility_mode && Z_TYPE_P(return_value) == IS_ARRAY)
+    {
+        int index = 0;
+        zval *zkey, *zvalue;
+        zval zret;
+        array_init(&zret);
+        
+        ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(z_arr), zkey)
+        {
+            zval zkey_str = *zkey;
```
Comment (twose) [inline swoole_redis_coro.cc]: you can use `zend::string`

File: swoole_redis_coro.cc [hunk 5]
```diff
@@ -3940,7 +3941,7 @@ static PHP_METHOD(swoole_redis_coro, hMGet)
     redis_request(redis, argc, argv, argvlen, return_value);
     SW_REDIS_COMMAND_FREE_ARGV
     
-    if(redis->compatibility_mode && Z_TYPE_P(return_value) == IS_ARRAY)
+    if (redis->compatibility_mode && Z_TYPE_P(return_value) == IS_ARRAY)
     {
         int index = 0;
```
Comment (twose) [inline swoole_redis_coro.cc]: `size_t` is better

File: swoole_redis_coro.cc [hunk 6]
```diff
@@ -1377,9 +1377,16 @@ static inline void sw_redis_command_key(INTERNAL_FUNCTION_PARAMETERS, const char
     SW_REDIS_COMMAND_ARGV_FILL(key, key_len)
     redis_request(redis, argc, argv, argvlen, return_value);
     
-    if (redis->compatibility_mode && memcmp("HGETALL", cmd, cmd_len) == 0 && Z_TYPE_P(return_value) == IS_ARRAY) 
+    if (redis->compatibility_mode)
     {
-        swoole_redis_handle_assoc_array_result(return_value, false);
+        if (Z_TYPE_P(return_value) == IS_ARRAY && memcmp("HGETALL", cmd, cmd_len) == 0)
```
Comment (twose) [inline swoole_redis_coro.cc]: you can use `ZVAL_IS_ARRAY`

**Comment by caohao-php** [comment]
Co\Redis Compatibility Mode 
--------
Co\Redis 的函数 hmGet/hGetAll/zrange/zrevrange/zrangebyscore/zrevrangebyscore 返回结果与php不一致的问题，已经得到解决 [#2529](https://github.com/swoole/swoole-src/pull/2529)。为了兼容老版本，在加上 $redis->setOptions(['compatibility_mode' => true]); 配置后，即可保证 Co\Redis 和 PHP Redis 返回结果一致。

```php
go(function() {
	$redis = new Swoole\Coroutine\Redis();
	$redis->setOptions(['compatibility_mode' => true]);
	$redis->connect(REDIS_SERVER_HOST, REDIS_SERVER_PORT);

	$co_get_val = $redis->get('novalue');
	$co_zrank_val = $redis->zRank('novalue', 1);
	$co_hgetall_val = $redis->hGetAll('hkey');
	$co_hmget_val = $redis->hmGet('hkey', array(3, 5));
	$co_zrange_val = $redis->zRange('zkey', 0, 99, true);
	$co_zrevrange_val = $redis->zRevRange('zkey', 0, 99, true);
	$co_zrangebyscore_val = $redis->zRangeByScore('zkey', 0, 99, ['withscores' => true]);
	$co_zrevrangebyscore_val = $redis->zRevRangeByScore('zkey', 99, 0, ['withscores' => true]);
});
```

**Pull request merged**